File: PracticasPDOORuby/Practica3_Ruby/lib/Irrgarten/player.rb
# frozen_string_literal: true
require_relative 'weapon'
require_relative 'shield'
require_relative 'dice'

module Irrgarten
  class Player
    # --- Constantes y Atributos de Lectura ---
    MAX_WEAPONS = 2
    MAX_SHIELDS = 3
    INITIAL_HEALTH = 10
    HITS2LOSE = 3

    attr_reader :row, :col, :name, :name, :intelligence, :strength, :health

    # ======================================================
    # MÉTODOS PÚBLICOS
    # ======================================================
    # Constructor
    # @param number Identificador del jugador (0, 1, 2, ...)
    # @param intelligence Nivel de inteligencia del jugador
    # @param strength Nivel de fuerza del jugador
    # @return Objeto de la clase Player
    def initialize(number, intelligence, strength)
      @number = number.to_s[0]
      @intelligence = intelligence.to_f
      @strength = strength.to_f
      @name = "Player#{@number}"
      @health = INITIAL_HEALTH.to_f
      @row = @col = -1
      @consecutive_hits = 0
      @weapons = []
      @shields = []
    end

    # Resucita al jugador, dejándolo como al principio
    # @return Nada (modifica el estado del jugador)
    def resurrect
      @weapons.clear
      @shields.clear
      @health = INITIAL_HEALTH
      @consecutive_hits = 0
    end

    # Establece la posición del jugador en el laberinto
    # @param row Fila de la posición
    # @param col Columna de la posición
    # @return Nada (modifica el estado del jugador)
    def set_pos(row, col) = (@row, @col = row.to_i, col.to_i)

    # Indica si el jugador está muerto
    # @return true si el jugador está muerto (salud <= 0), false en caso contrario
    def dead? = @health <= 0

    # Calcula el ataque del jugador (fuerza + armas)
    # @return Valor del ataque del jugador
    def attack
      @strength + sum_weapons
    end

    # Defiende al jugador de un ataque recibido
    # @param received_attack Valor del ataque recibido
    # @return Nada (modifica el estado del jugador)
    def defend(received_attack)
      manage_hit(received_attack)
    end

    # Representación en cadena del jugador
    # @return Cadena con la información del jugador
    def to_s
      "Player{nombre = '#{@name}', inteligencia = '#{@intelligence}', fuerza = '#{@strength}', salud = '#{@health}', posicion = (#{@row}, #{@col}), armas = '#{@weapons}', escudos = '#{@shields}'"
    end

    # ======================================================
    # P3 - Métodos públicos a implementar
    # ======================================================

    # Decide la dirección final del movimiento del jugador.
    # Si la dirección preferida es válida, se devuelve.
    # Si no es válida, pero hay movimientos válidos, se devuelve el primero de la lista de válidos.
    # Si no hay movimientos válidos, se devuelve la preferida aunque sea inválida.
    # @param direction
    # @param valid_moves
    # @return Dirección final
    def move(direction, valid_moves)
      size = valid_moves.length # Diagrama 1.1: size()
      contained = valid_moves.include?(direction) # Diagrama 1.2: contains(direction)

      # Diagrama alt: [(size > 0) && (!contained)]
      if size > 0 && !contained
        @first_element = valid_moves[0] # Diagrama 1.3: get(0)

        return @first_element # Diagrama 1.4: return first_element
      else
         direction # Diagrama 1.5: return direction
      end
    end

    # Recibe las recompensas (armas, escudo y salud) al ganar un combate.
    # Delega en Dice para obtener las cantidades y en los métodos privados
    # receive_weapon y receive_shield para gestionarles
    def receive_reward
      w_reward = Dice.weapons_reward # Diagrama 1.1: weapons_reward
      s_reward = Dice.shields_reward # Diagrama 1.2: shields_reward

      # Diagrama loop: [1, wReward]
      w_reward.times do
        wnew = new_weapon # Diagrama 1.3: new_weapon

        receive_weapon(wnew) # Diagrama 1.4: receive_weapon(wnew)
      end

      # Diagrama loop: [1, sReward]
      s_reward.times do
        snew = new_shield # Diagrama 1.5: new_shield

        receive_shield(snew) # Diagrama 1.6: receive_shield(snew)
      end

      extra_health = Dice.health_reward # Diagrama 1.7: health_reward

      @health += extra_health # Diagrama: health += extra_health
    end

    # =====================================================
    # MÉTODOS PRIVADOS
    # =====================================================
    private

    # Calcula la energía defensiva del jugador (inteligencia + escudos)
    # @return Valor de la energía defensiva del jugador
    def defensive_energy
      @intelligence + sum_shields
    end

    # Resetea el contador de golpes consecutivos
    # @return Nada (modifica el estado del jugador)
    def reset_hits
      @consecutive_hits = 0
    end

    # Incrementa el contador de golpes consecutivos
    # @return Nada (modifica el estado del jugador)
    def inc_consecutive_hits
      @consecutive_hits += 1
    end

    # Suma el poder de ataque de todas las armas del jugador
    # @return Suma del poder de ataque de las armas
    def sum_weapons
      @weapons.sum {|w| w.attack}
    end

    # Suma la protección de todos los escudos del jugador
    # @return Suma de la protección de los escudos
    def sum_shields
      @shields.sum {|s| s.protect}
    end

    # Crea un nuevo arma con poder y usos aleatorios
    # @return Nuevo objeto de la clase Weapon
    def new_weapon
      Weapon.new(Dice.weapon_power, Dice.uses_left)
    end

    # Crea un nuevo escudo con poder y usos aleatorios
    # @return Nuevo objeto de la clase Shield
    def new_shield
      Shield.new(Dice.shields_power, Dice.uses_left)
    end

    # El jugador ha sido herido (pierde 1 punto de salud)
    # @return Nada (modifica el estado del jugador)
    def got_wounded
      @health -= 1
    end

    # ======================================================
    # P3 - Métodos privados a implementar
    # ======================================================

    # Gestiona la recepción de un arma.
    # Itera sobre las armas actuales, descartando (eliminando) las que
    # ya no tengan usos. Si tras la limpieza hay espacio (< MAX_WEAPONS), se aniade la nueva arma.
    # @param w
    def receive_weapon(w)
      # Diagrama loop: [for all] + 1.1 + 1.2 + 1.3
      # delete_if itera y elimina si el bloque es true
      @weapons.delete_if { |wi| wi.discard }

      size = @weapons.length # Diagrama 1.4: size()

      # Diagrama opt: [size < MAX_WEAPONS]
      if (size < MAX_WEAPONS)
        @weapons << w # Diagrama 1.5: add(w)
      end
    end

    # Gestiona la recepción de un escudo.
    # Itera sobre los escudos actuales, descartando (eliminando) los que
    # ya no tengan usos. Si tras la limpieza hay espacio (< MAX_SHIELDS), se aniade el nuevo escudo.
    # @param s
    def receive_shield(s)
      # Diagrama loop: [for all] + 1.1 + 1.2 + 1.3
      # delete_if itera y elimina si el bloque es true
      @shields.delete_if { |si| si.discard }

      size = @shields.length # Diagrama 1.4: size()

      # Diagrama opt: [size < MAX_SHIELDS]
      if (size < MAX_SHIELDS)
        @shields << s # Diagrama 1.5: add(s)
      end
    end

    # Gestiona la lógica de un impacto recibido.
    # Compara la defensa del jugador con el ataque. Si es superado,
    # recibe daño y acumula impactos. Si resiste, reinicia impactos.
    # Comprueba si el jugador pierde (por vida o por impactos consecutivos).
    # @param received_attack
    # @return true si el jugador pierde, false en caso contrario
    def manage_hit(received_attack)
      defense = defensive_energy # Diagrama 1.1: defensive_energy()

      # Diagrama alt: [defense < received_attack]
      if defense < received_attack
        got_wounded # Diagrama 1.2: got_wounded()
        inc_consecutive_hits # Diagrama 1.3: inc_consecutive_hits()
      else
        reset_hits # Diagrama 1.4: reset_hits()
      end

      # Diagrama alt: [( (consecutive_hits >= HITS2LOSE) || (dead?))]
      if (@consecutive_hits >= HITS2LOSE) || dead?
        reset_hits # Diagrama 1.5: reset_hits()
        lose = true # Diagrama: lose = true
      else
        lose = false # Diagrama: lose = false
      end

      lose # Diagrama 2: return lose
    end
  end
end

File: PracticasPDOOJava/Practica5_Java/src/irrgarten/Player.java
package irrgarten;

import java.util.ArrayList;
import java.util.Iterator; // Nueva implementación

/**
 * Representa un jugador dentro del laberinto
 * <p>
 * Cada jugador posee características propias como inteligencia, fuerza y salud,
 * además de una posición (row/column) dentro del {@link labyrinth}.
 * Puede equipar un número de limitado de armas y escudos, atacar, defenderse y 
 * resucitar tras ser derrotado.
 * </p>
 * 
 * @author Joaquín Cruz Lorenzo
 */
public class Player extends LabyrinthCharacter {
    // ==================================
    // CONSTANTES DE CLASE
    // ==================================
    private static final int MAX_WEAPONS = 2;
    private static final int MAX_SHIELDS = 3;
    private static final int INITIAL_HEALTH = 10;
    private static final int HITS2LOSE = 3;
    
    private char number;
    private int consecutiveHits = 0;
    
    private ArrayList<Weapon> weapons = new ArrayList();
    private ArrayList<Shield> shields = new ArrayList();
    
    // ELIMINO todos los atributos de instancia repetidos. 
    // Ya que los usamos en LabyrinthCharacter
    // ==================================
    // ATRIBUTOS DE INSTANCIA 
    // ==================================
    // private String name; 
    // private float intelligence, strength, health;
    // private int row, col, 
    
    // ===============================================
    // MÉTODOS PÚBLICOS
    // ===============================================
    
    /**
     * Constructor principal del jugador.
     * @param number
     * @param intelligence
     * @param strength 
     */
    public Player(char number, float intelligence, float strength) {
        super ("Player " + number, intelligence, strength, INITIAL_HEALTH);
        this.number = number;
    }
    
    public Player(Player other) {
        // Llamo al constructor de copia de la superclase (LabyrinthCharacter)
        super(other);
        
        // Copiamos los atributos PROPIOS de Player
        this.number = other.number;
        this.consecutiveHits = other.consecutiveHits;
        
        // Copia de armas y escudos
        for (Weapon w : weapons) {
            this.weapons.add(new Weapon(w));
        }
        
        for (Shield s : shields) {
            this.shields.add(new Shield(s));
        }
    }
    
    /**
     * Restaura al jugador al estado inicial.
     * <ul>
     *          <li> Reinicia salud e impactos consecutivos. </li>
     * </ul>
     */
    public void resurrect() {
        setHealth(INITIAL_HEALTH); // setHealth es protected de LabyrinthCharacter.
        this.consecutiveHits = 0;
    }

    public char getNumber() { return number; }

    
    
    /**
     * Calucla la potencia de ataque del jugador.
     * <p> La fórmula básica es {@code strength + sumWeapons()} </p>
     * @return  potencia de ataque total
     */
    public float attack() {
        return getStrength() + sumWeapons();
    }
    
    /**
     * Inicia la defensa ante un ataque recibido.
     * <p> Lo dejamos sin implementar en P2. Se nos dará información en P3 </p>
     * @param receivedAttack intensidad del ataque recibido
     * @return 
     */
    public boolean defend(float receivedAttack) {
        // Diagrama 1.1: manageHit(received attack)
        return manageHit(receivedAttack);
    }
    
    @Override
    public String toString() {
        String baseState = super.toString();
        String playerState = "Hits: " + consecutiveHits + "\n" + " Weapons: " + weapons.toString() + "\n" + " Shields: " + shields.toString();
        
        return baseState + playerState;
    }
    
    // ==================================
    // P3 - Métodos públicos a implementar
    // ==================================
    /**
     * Decide la dirección final del movimiento del jugador
     * <p>
     * Si la dirección preferida es válida, se devuelve.
     * Si no lo es, pero hay movimientos válidos, se devuelve el primero de la lista de no válidos
     * Si no hay movimientos válidos, se devuelve la preferida (aunque sea inválida)
     * </p>
     * @param direction Dirección preferida
     * @param validMoves Array de direcciones válidas ñ
     * @return Dirección final
     */
    public Directions move(Directions direction, Directions[] validMoves){ 
        int size = validMoves.length; // Diagrama 1.1: size()
        
        // Diagrama 1.2: contains(directions)
        boolean contained = false; 
        for (Directions d : validMoves) {
            if (d == direction) {
                contained = true;
                break;
            }
        }
        
        // Diagrama alt: [(size > 0) && (!contained)]
        if (size > 0 && !contained) {
            // Diagrama 1.3: get(0)
            // Diagrama 1.4: firstElement
            return validMoves[0];
        } else {
            // Diagrama 1.5: direction (return)
            return direction;
        }
    }
    
    /**
     * Recibe las recompensas (armas, escudos, salud) al ganar un combate. 
     * Delega en Dice para obtener las cantidades y en los métodos privados 
     * receiveWeapon y receiveShield para gestionarlas
     */
    public void receiveReward() { 
        int wReward = Dice.weaponsReward(); // Diagrama 1.1: weaponsReward
        int sReward = Dice.shieldsReward(); // Diagrama 1.2: shieldsReward
        
        // Diagrama loop: (1, wReward)
        for (int i = 1; i <= wReward; i++){
            Weapon wnew = newWeapon(); // Diagrama 1.3: newWeapon()
            receiveWeapon(wnew); // Diagrama 1.4: receiveWeapon(w: Weapon = wnew)
        }
        
        // Diagrama loop: (1, sReward)
        for (int i = 1; i <= sReward; i++){
            Shield snew = newShield(); // Diagrama 1.5: newShield()
            receiveShield(snew); // Diagrama 1.6: receiveShield(s: Shield = snew)
        }
        
        int extraHealth = Dice.healthReward(); // Diagrama 1.7: healthReward
        
       
        setHealth(getHealth() + extraHealth);
    }
    
    
    
    // ==========================================================
    // MÉTODOS PRIVADOS
    // ==========================================================
    /**
     * Suma la potencia de todas las armas equipadas (consumiendo sus usos)
     * @return suma total de ataque de armas activas
     */
    protected float sumWeapons() {
        float total = 0.0f;
        
        for (Weapon w : weapons) total += w.attack();
        
        return total;
    }
    
    /**
     * Suma la capacidad de protección de todos los escudos (consumiendo sus usos)
     * @return  protección total ofrecida por los escudos
     */
    protected float sumShields() {
        float total = 0.0f;
        
        for (Shield s : shields) total += s.protect();
        
        return total;
    }
    
    /**
     * Calcula la energía defensiva del jugador.
     * <p> {@code defensiveEnergy = intelligence + sumShields()} </p>
     * @return  valor de energía defensiva
     */
    protected float defensiveEnergy() {
        return getIntelligence() + sumShields();
    }
    
    /**
     * Incrementa el número de impactos consecutivos sufridos.
     */
    private void incConsecutiveHits() {
        consecutiveHits++;
    }
    
    /**
     * Reinicia el contador de impactos consecutivos a 0.
     */
    private void resetHits() {
        consecutiveHits = 0;
    }
    

    
    /**
     * Genera un arma nueva con valores aleatorios dados por {@link Dice}.
     * @return  nueva instancia de {@link Weapon}
     */
    private Weapon newWeapon() {
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }
    
    /**
     * Genera un escudo nuevo con valores aleatorios dados por {@link Dice}.
     * @return nueva instancia de {@link Shield}
     */
    private Shield newShield() {
        return new Shield(Dice.shieldPower(), Dice.usesLeft());
    }
    
    // =========================================
    // P3 - Métodos privados a implementar
    // =========================================
    /**
     *  Gestiona la recepción de un arma.
     * <p>
     * Se itera sobre las armas actuales, descartando (eliminando) las que 
     * ya no tengan usos. Si tras la limpia hay espacio (< MAX_WEAPONS),
     * se aniade la nueva arma.
     * </p>
     * @param w Arma a recibir
     */
    private void receiveWeapon(Weapon w) {
        // Diagrama: loop for all
        Iterator<Weapon> it = weapons.iterator();
        
        while (it.hasNext()){
            Weapon wi = it.next(); // Diagrama 1.1: next()
            
            // Diagrama opt [discard]
            if (wi.discard()) { // Diagrama 1.2: discard()
                 it.remove(); // Diagrama 1.3: remove(wi)
            }
        }
        
        // Diagrama 1.4: int size()
        int size = weapons.size();
        
        // Diagrama opt [size < MAX_WEAPONS]
        if(size < MAX_WEAPONS){
            weapons.add(w); // Diagrama 1.5: add(w)
        }
    }
    
    /**
     * Gestiona la recepción de un escudo.
     * <p> 
     * Se itera sobre los escudos actuales, descartando (eliminando) los que ya no tengan usos.
     * Si tras la limpieza hay espacio (<   MAX_SHIELDS), se aniade el nuevo escudo.
     * </p>
     * @param s Escudo a recibir. 
     */
    private void receiveShield(Shield s) { 
        // Diagrama: loop [for all]
        Iterator<Shield> it = shields.iterator();
        
        while(it.hasNext()) {
            Shield si = it.next(); // Diagrama 1.1: next()
            
            // Diagrama opt [discard]
            if(si.discard()){ // Diagrama 1.2: discard()
                it.remove(); // Diagrama 1.3: remove()
            }
        }
        
        // Diagrama 1.4: size()
        int size = shields.size();
        
        // Diagrama opt [size < MAX_SHIELDS]
        if (size < MAX_SHIELDS){
            shields.add(s); // Diagrama 1.5: add(s)
        }
    }
    
    /**
     * Gestiona la lógica de un impacto recibido.
     * <p>
     * Compara la defensa del jugador con el ataque. Si es superado, 
     * recibe danio y acumula impactos. Si resiste, reinicia impactos.
     * Comprueba si el jugador pierde (por vida o por impactos consecutivos)
     * </p>
     * @param receivedAttack Ataque recibido
     * @return true si el jugador pierde, false si sobrevive
     */
    private boolean manageHit(float receivedAttack) { 
         // Diagrama 1.1.1: defensiveEnergy()
        float defense = defensiveEnergy();
        boolean lose = false; // Variable local para gestionar el resultado.
        
        // Diagrama alt: [defense < receivedAttack]
        if (defense < receivedAttack){
            gotWounded() ; // Diagrama 1.1.2: gotWounded()
            incConsecutiveHits(); // Diagrama 1.1.3: incConsecutiveHits()
        } else {
            resetHits();
        }
        
        // Diagrama alt: [((consecutiveHits == HITS2LOSE) || (dead()))]
        if ( (consecutiveHits == HITS2LOSE) || (dead()) ) {
            resetHits(); // Diagrama 1.1.5: resetHits()
            
            // Diagrama lose = true
            lose = true;
        } 
        
        return lose;
    }
    
    // LISTA DE MÉTODOS SUPRIMIDOS YA QUE SE USAN AHORA EN LabyrinthCharacter
    // public int getRow() { return row; }
    // public int getCol() { return col; }
    // public char getNumber() { return number; }
    // public void setPos(int row, int col) {
    //    this.row = row;
    //    this.col = col;
    // }
    // private void gotWounded() { health--; }
    // public boolean dead() { return health <= 0; }
}
